### PR TITLE
SLCORE-67 fetch and store issues

### PR DESCRIPTION
**Pull request by janos-ss**

### CODE REVIEW

**Comment by janos-ss** [comment]
@dbmeneses A bit experimental, WIP... Let me know what you think so far.

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/IssueDownloader.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.container.connected.update;
+
+import org.sonar.scanner.protocol.input.ScannerInput;
+
+public interface IssueDownloader {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/IssueDownloader.java]: CRITICAL Annotate the "IssueDownloader" interface with the @FunctionalInterface annotation [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1609)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/IssueStoreFactory.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.container.connected.update;
+
+import java.nio.file.Path;
+
+public interface IssueStoreFactory {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/IssueStoreFactory.java]: CRITICAL Annotate the "IssueStoreFactory" interface with the @FunctionalInterface annotation [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1609)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* CRITICAL 2 critical

Watch the comments in this conversation to review them.

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/IssueDownloader.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.container.connected.update;
+
+import org.sonar.scanner.protocol.input.ScannerInput;
+
+@FunctionalInterface
+public interface IssueDownloader {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/IssueDownloader.java]: MAJOR Drop this interface in favor of "java.util.function.Function<String,Iterable>". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1711)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/IssueStoreFactory.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.container.connected.update;
+
+import java.nio.file.Path;
+
+@FunctionalInterface
+public interface IssueStoreFactory {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/IssueStoreFactory.java]: MAJOR Drop this interface in favor of "java.util.function.Function<Path,IssueStore>". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1711)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/IssueUtils.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import org.sonar.scanner.protocol.input.ScannerInput;
+
+public class IssueUtils {
+  private IssueUtils() {
+    // utility class, forbidden constructor
+  }
+
+  public static Map<String, Iterable<ScannerInput.ServerIssue>> groupByFileKey(Iterable<ScannerInput.ServerIssue> issues) {
+    Map<String, Iterable<ScannerInput.ServerIssue>> groupedIssues = new HashMap<>();
+    for (ScannerInput.ServerIssue issue : issues) {
```
Comment (dbmeneses) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/IssueUtils.java]: something like issues.stream().collect(Collectors.groupingBy(IssueUtils::createFileKey) could be easier?

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/InMemoryIssueStore.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.container.connected.update;
+
+import java.util.Collections;
+import java.util.Map;
+import org.sonar.scanner.protocol.input.ScannerInput;
+
+public class InMemoryIssueStore implements IssueStore {
```
Comment (dbmeneses) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/InMemoryIssueStore.java]: Should not be kept in memory in the final implementation

Comment (janos-ss) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/InMemoryIssueStore.java]: Added `TODO` here and at the place it's currently used. I have a disk-based implementation in progress.

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/InMemoryIssueStore.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.container.connected.update;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import org.sonar.scanner.protocol.input.ScannerInput;
+
+// TODO for testing only. Move to /test/ at the soonest
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/InMemoryIssueStore.java]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleConfigUpdateExecutor.java [hunk 8]
```diff
@@ -42,22 +45,50 @@
...
 public class ModuleConfigUpdateExecutor {
 
   private final StorageManager storageManager;
   private final SonarLintWsClient wsClient;
+  private final IssueDownloader issueDownloader;
+  private final IssueStoreFactory issueStoreFactory;
   private final TempFolder tempFolder;
 
   public ModuleConfigUpdateExecutor(StorageManager storageManager, SonarLintWsClient wsClient, TempFolder tempFolder) {
+    // TODO replace InMemoryIssueStore with persistent (filesystem-based) implementation (in progress)
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleConfigUpdateExecutor.java]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* INFO 2 info

Watch the comments in this conversation to review them.

**Pull request merged**